tools/ctexplain/bazel_api.py

```diff
@@ -0,0 +1,161 @@
+import os
+import re
+import subprocess
+from tools.ctexplain.types import Configuration
+from tools.ctexplain.types import ConfiguredTarget
+from tools.ctexplain.types import HostConfiguration
+from tools.ctexplain.types import NullConfiguration
+
+
+def run_bazel_in_client(args):
+  """Calls bazel within the current workspace.
+
+  For production use. Tests use an alternative invoker that goes through test
+  infrastructure.
+
+  Args:
+    args: the arguments to call Bazel with
+
+  Returns:
+    Tuple of (return code, stdout, stderr)
+  """
+  result = subprocess.run(
+      ["bazel"] + args,
+      cwd=os.getcwd(),
+      stdout=subprocess.PIPE,
+      stderr=subprocess.PIPE,
+      check=True)
+  return [
+      result.returncode,
+      result.stdout.decode("utf-8").split(os.linesep), result.stderr
+  ]
+
+
+class BazelApi():
+  """API that accepts injectable Bazel invocation logic."""
+
+  def __init__(self, run_bazel=run_bazel_in_client):
+    self.run_bazel = run_bazel
+
+  def cquery(self, args):
+    """Calls cquery with the given arguments.
+
+    Args:
+      args: A list of cquery command-line arguments, one argument per entry.
+
+    Returns:
+      (success: bool, stderr: str, cts: Tuple[ConfiguredTarget]), where
+      success is True iff the query succeeded, stderr contains the query's
+      stderr (regardless of success value), and cts is the configured targets
+      found by the query if successful, empty otherwise.
+    """
+    base_args = ["cquery", "--show_config_fragments=transitive"]
+    (returncode, stdout, stderr) = self.run_bazel(base_args + args)
+    if returncode != 0:
+      return (False, stderr, ())
+
+    cts = set()
+    for line in stdout:
+      ctinfo = _parse_cquery_result_line(line)
+      if ctinfo is not None:
+        cts.add(ctinfo)
+
+    return (True, stderr, tuple(cts))
+
+  def get_config(self, config_hash):
+    """Calls "bazel config" with the given config hash.
+
+    Args:
+      config_hash: A config hash as reported by "bazel cquery".
+
+    Returns:
+      A types.Configuration with the matching configuration or None if no match
+      is found.
+
+    Raises:
+      ValueError on any parsing problems.
+    """
+    if config_hash == "HOST":
+      return HostConfiguration()
+    elif config_hash == "null":
+      return NullConfiguration()
+
+    base_args = ["config", "--output=json"]
+    (returncode, stdout, stderr) = self.run_bazel(base_args + [config_hash])
+    if returncode != 0:
+      raise ValueError("Could not get config: " + stderr)
+    config_json = json.loads(os.linesep.join(stdout))
+    fragments = [
+        fragment["name"].split(".")[-1] for fragment in config_json["fragments"]
+    ]
+    options = {
+        entry["name"].split(".")[-1]: entry["options"]
+        for entry in config_json["fragmentOptions"]
+    }
+    return Configuration(fragments, options)
+
+
+##############################################
+# Regex patterns for matching cquery results:
```

Comment:
I am obligated to ask if you tried this matching using non-regex Python string manipulation/search functions first.

I am also a little concerned that the capturing groups end up nested inside some of the substrings to build up the regex but not all.

Reply:
Done. Less code, too!